File: src/domain/models/OrderItem.ts
// ============================================================
// OrderItem Value Object - 訂單項目值物件
// ============================================================
// 不可變的值物件，封裝了訂單中每個商品的資訊。
// 遵循 SRP：只負責表達「訂單中的一項商品」這個概念。
// ============================================================

import { InvalidOrderItemError } from '../errors/DomainError';

export class OrderItem {
  readonly subtotal: number;

  constructor(
    readonly productId: string,
    readonly productName: string,
    readonly unitPrice: number,
    readonly quantity: number,
  ) {
    if (!productId || productId.trim() === '') {
      throw new InvalidOrderItemError('產品 ID 不能為空');
    }
    if (quantity <= 0) {
      throw new InvalidOrderItemError('數量必須大於 0');
    }
    if (unitPrice < 0) {
      throw new InvalidOrderItemError('單價不能為負數');
    }
    this.subtotal = unitPrice * quantity;
  }
}

File: src/main/java/com/example/hexagonal/domain/model/OrderItem.java
package com.example.hexagonal.domain.model;

// ============================================================
// OrderItem Value Object - 訂單項目值物件
// ============================================================
// 使用 Java 21 Record，不可變。
// 遵循 SRP：只負責表達「訂單中的一項商品」概念。
// ============================================================

import com.example.hexagonal.domain.exception.InvalidOrderItemException;

public record OrderItem(
        String productId,
        String productName,
        double unitPrice,
        int quantity
) {
    public OrderItem {
        if (productId == null || productId.isBlank()) {
            throw new InvalidOrderItemException("產品 ID 不能為空");
        }
        if (quantity <= 0) {
            throw new InvalidOrderItemException("數量必須大於 0");
        }
        if (unitPrice < 0) {
            throw new InvalidOrderItemException("單價不能為負數");
        }
    }

    public double subtotal() {
        return unitPrice * quantity;
    }
}
